### PR TITLE
feat(membership-plans): improvements 

### PR DESCRIPTION
**Pull request by adekbadek**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Some tweaks to the Membership Plans table:

1. Wording changes (e.g. "Active Members" -> "Active Memberships")
2. Adds an "Active Subscriptions" column:

image

3. Enables table sorting by Network ID
4. Removes "Links" column in favour of linking with the plan name

(Note: this is a part of integrating the changes from the `data-integrity-improvements` branch (https://github.com/Automattic/newspack-network/pull/89))

### How to test the changes in this Pull Request:

1. Set `NEWSPACK_NETWORK_EXPERIMENTAL_AUDITING_FEATURES` env variable to `true`
1. Visit the "Membership Plans" view on the Hub site and observe the changes listed above

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Review by chickenn00dle — COMMENTED** [review]
I tried following the testing steps but was only able to confirm the second two points. I am not seeing the `Active Memberships` or `Active Subscriptions` columns though:

Screenshot 2024-07-12 at 11 13 56

Am I missing a step to test this?

**Comment by adekbadek** [comment]
Yes, sorry! I've update the testing step. What's missing is the `NEWSPACK_NETWORK_EXPERIMENTAL_AUDITING_FEATURES` env variable.

File: includes/woocommerce-memberships/class-admin.php [hunk 1]
```diff
@@ -100,14 +101,45 @@ function ( $membership ) {
...
+	 * Filter user membership data from REST API.
+	 *
+	 * @param \WP_REST_Response $response the response object.
+	 * @param null|\WP_Post     $user the user membership post object.
+	 * @param \WP_REST_Request  $request the request object.
+	 */
+	public static function add_data_to_wc_user_membership_response( $response, $user, $request ) {
+		if ( $request && isset( $request->get_headers()['x_np_network_signature'] ) ) {
+			// Add network plan ID to the response.
+			$plan = wc_memberships_get_membership_plan( $response->data['plan_id'] );
```
Comment (chickenn00dle) [inline includes/woocommerce-memberships/class-admin.php:137]: Looks like [`wc_memberships_get_membership_plan`](https://woocommerce.com/document/woocommerce-memberships-function-reference/#section-6) can return `false`, so we will probably need to check the value of `$plan` before adding to response data.

File: includes/hub/admin/class-membership-plans.php [hunk 2]
```diff
@@ -104,27 +103,29 @@ private static function get_membership_plans_from_cache() {
...
 
 		$nodes = \Newspack_Network\Hub\Nodes::get_all_nodes();
 		foreach ( $nodes as $node ) {
-			$node_plans = self::fetch_collection_from_api( $node, 'wc/v2/memberships/plans', 'membership-plans' );
+			$query_args = [];
+			if ( \Newspack_Network\Admin::use_experimental_auditing_features() ) {
+				$query_args['include_active_members_emails'] = 1;
+			}
+			$node_plans = self::fetch_collection_from_api( $node, 'wc/v2/memberships/plans', 'membership-plans', $query_args );
 			foreach ( $node_plans as $plan ) {
```
Comment (chickenn00dle) [inline includes/hub/admin/class-membership-plans.php:132]: Getting the following warning on this line:

```
PHP Warning:  foreach() argument must be of type array|object, null given in /Users/raz/Sites/newspack/network/includes/hub/admin/class-membership-plans.php on line 129
```

File: includes/hub/admin/class-membership-plans.php [hunk 3]
```diff
@@ -75,16 +75,15 @@ public static function render() {
...
-		if ( Network_Admin::use_experimental_auditing_features() ) {
-			$endpoint = add_query_arg( 'include_active_members_emails', 1, $endpoint );
-		}
+	public static function fetch_collection_from_api( $node, $collection_endpoint, $collection_endpoint_id, $query_args = [] ) {
+		$endpoint = add_query_arg( $query_args, sprintf( '%s/wp-json/%s', $node->get_url(), $collection_endpoint ) );
 		$response = wp_remote_get( // phpcs:ignore
 			$endpoint,
 			[
 				'headers' => $node->get_authorization_headers( 'get-woo-' . $collection_endpoint_id ),
+				'timeout' => 60, // phpcs:ignore
```
Comment (chickenn00dle) [inline includes/hub/admin/class-membership-plans.php:86]: nonblocking nit: Does the default not work here?

File: includes/hub/admin/class-membership-plans-table.php [hunk 4]
```diff
@@ -22,25 +22,54 @@ class Membership_Plans_Table extends \WP_List_Table {
...
 	 * Prepare items to be displayed
 	 */
 	public function prepare_items() {
-		$this->_column_headers = [ $this->get_columns(), [], [], 'id' ];
-		$this->items = Membership_Plans::get_membershp_plans_from_network();
+		$membership_plans_from_network_data = Membership_Plans::get_membership_plans_from_network();
+
+		// Handle table sorting.
+		$order = $_REQUEST['order'] ?? false; // phpcs:ignore WordPress.Security.NonceVerification.Missing, WordPress.Security.NonceVerification.Recommended, WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
+		$orderby = $_REQUEST['orderby'] ?? false; // phpcs:ignore WordPress.Security.NonceVerification.Missing, WordPress.Security.NonceVerification.Recommended, WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
```
Comment (chickenn00dle) [inline includes/hub/admin/class-membership-plans-table.php]: Can we sanitize these to address the `WordPress.Security.ValidatedSanitizedInput.InputNotSanitized` warning?

**Review by chickenn00dle — CHANGES_REQUESTED** [review]
Thanks @adekbadek!

Tested again and it looks to be working as expected 👍 

I left a few small bits of feedback and questions, then this should be good to go.

File: includes/hub/admin/class-membership-plans.php (hunk 3 above)
Comment (adekbadek) [reply includes/hub/admin/class-membership-plans.php:86]: What do you mean? Default timeout is `5`.

File: includes/hub/admin/class-membership-plans-table.php (hunk 4 above)
Comment (adekbadek) [reply includes/hub/admin/class-membership-plans-table.php]: Done in a4ab6df816d0fa3747ff8f5a236c56a5e7653afe

File: includes/hub/admin/class-membership-plans.php (hunk 2 above)
Comment (adekbadek) [reply includes/hub/admin/class-membership-plans.php:132]: Fixed in a4ab6df816d0fa3747ff8f5a236c56a5e7653afe

File: includes/woocommerce-memberships/class-admin.php (hunk 1 above)
Comment (adekbadek) [reply includes/woocommerce-memberships/class-admin.php:137]: Fixed in a4ab6df816d0fa3747ff8f5a236c56a5e7653afe

File: includes/hub/admin/class-membership-plans.php (hunk 3 above)
Comment (chickenn00dle) [reply includes/hub/admin/class-membership-plans.php:86]: Ah. Whoops. For some reason I thought this was `30`. Nevermind!

**Review by chickenn00dle — APPROVED** [review]

**Pull request merged**